framework/DataObjects/DataSet.py

```diff
@@ -0,0 +1,1853 @@
+    """
+    self._data = None
+    self._collector = None
+    self._meta = {}
+    # TODO others?
+    self._alignedIndexes = {}
+    self._scaleFactors = {}
+
+  def sliceByIndex(self,index):
+    """
+      Returns list of realizations at "snapshots" along dimension "index".
+      For example, if 'index' is 'time', then returns cross-sectional slices of the dataobject at each recorded 'time' index value.
+      @ In, index, str, name of index along which to obtain slices
+      @ Out, slices, list, list of xr.Dataset slices.
+    """
+    data = self.asDataset()
+    # if empty, nothing to do
+    if self._data is None or len(self._data) == 0:
+      self.raiseAWarning('Tried to return sliced data, but DataObject is empty!')
+      return []
+    # assert that index is indeed an index
+    if index not in self.indexes + [self.sampleTag]:
+      self.raiseAnError(IOError,'Requested slices along "{}" but that variable is not an index!  Options are: {}'.format(index,self.indexes))
+    numIndexCoords = len(data[index])
+    slices = list(data.isel(**{index:i}) for i in range(numIndexCoords))
+    # NOTE: The slice may include NaN if a variable does not have a value along a different index for this snapshot along "index"
+    return slices
+
+  def write(self,fileName,style='netCDF',**kwargs):
+    """
+      Writes this dataset to disk based on the format.
+      @ In, fileName, str, path and name of file to write
+      @ In, style, str, optional, options are enumerated below
+      @ In, kwargs, dict, optional, additional arguments to pass to writing function
+          Includes:  firstIndex, int, optional, if included then is the realization index that writing should start from (implies appending instead of rewriting)
+      @ Out, index, int, index of latest rlz to be written, for tracking purposes
+    """
+    self.asDataset() #just in case there is stuff left in the collector
+    if style.lower() == 'netcdf':
+      self._toNetCDF(fileName,**kwargs)
+    elif style.lower() == 'csv':
+      if len(self._data[self.sampleTag])==0: #TODO what if it's just metadata?
+        self.raiseAWarning('Nothing to write!')
+        return
+      #first write the CSV
+      firstIndex = kwargs.get('firstIndex',0)
+      self._toCSV(fileName,start=firstIndex,**kwargs)
+      # then the metaxml
+      self._toCSVXML(fileName,**kwargs)
+    # TODO dask?
+    else:
+      self.raiseAnError(NotImplementedError,'Unrecognized write style: "{}"'.format(style))
+    if not self.hierarchical and 'RAVEN_isEnding' in self.getVars():
+      return len(self._data.where(self._data['RAVEN_isEnding']==True,drop=True)['RAVEN_isEnding'])
+    else:
+      return len(self) # so that other entities can track which realization we've written
+
+  ### INITIALIZATION ###
+  # These are the necessary functions to construct and initialize this data object
+  def __init__(self):#, in_vars, out_vars, meta_vars=None, dynamic=False, var_dims=None,cacheSize=100,prealloc=False):
```

Comment:
Having a dunder init this low in a class definition is rather unconventional.

Reply:
I've never been 100% happy with it, either.   I really like the organization of the API for the data object for developers, since only very rarely will a developer need to create a data object instance for themselves; however, it does put the constructor in an atypical location.  Thoughts?

Reply:
Anyone reading your code with expect init to be at the top of a class definition and know to ignore it if they don't need it. If they don't see it there, they will assume it doesn't have one, possibly creating a source of confusion. I vote move it to the top.

Reply:
moved on consensus.